### PR TITLE
fix bug on removing support

### PR DESCRIPTION
**Pull request by Qianfinland**

### CODE REVIEW

**Comment by oliverbarnes** [comment]
Cool, just waiting on the tests as discussed on the github issue for the bug. Maybe also add an alert to the user when she clicks on removing support, warning that she won't be able to post suggestions, and if she has suggestions already posted, they won't be shown anymore

**Comment by Qianfinland** [comment]
hi Oliver, currently we haven't associated suggestions with users, how could i implement when remove support then also remove suggestions the users had typed?

File: app/routes/initiative.js [hunk 1]
```diff
@@ -9,12 +9,20 @@ var InitiativeRoute = Ember.Route.extend({
...
+      initiative.set('isSupported', true);
     },
 
     removeSupport: function() {
       var initiative = this.get('controller.model');
-      initiative.set('isSupported', false)
+      initiative.set('isSupported', false);
+      this.transitionTo('initiative');
+      // make pop-up alert when click link on remove support but don't yet find how to test pop-up window
+      // if (confirm("You cannot make suggestion anymore, are you sure?")){
```
Comment (oliverbarnes) [inline app/routes/initiative.js]: "Are you sure you'd like to remove support? You won't be able to make suggestions after that, and any suggestions made will be removed"

**Comment by oliverbarnes** [comment]
Good question - let's leave it for when we do have users implemented, please add a feature request here on github for it?

**Comment by Qianfinland** [comment]
about the pop-up alert  feature, i still didn't figure out how to write test for it. As you can see i just add the confirm code in inside of the removeSupport action, the alert message is not in the template, i don't know how to test it. Also there is cancel and ok button for users to click, i have no idea how to write test. Any suggestion?

**Comment by Qianfinland** [comment]
about the adding feature request, do you mean i make a issue title
something like "removing suggestions when remove support" on github? I can
make issue but i cannot find how to label the issue as feature, bug. I
guess i don't right to label it as i am not owner. Correct me if i am wrong

**Comment by oliverbarnes** [comment]
For the pop-up, you'd have to find it with jquery somehow (I don't know from the top of my head). But googling a bit I found [this cookbook entry for ember modals](http://emberjs.com/guides/cookbook/user_interface_and_interaction/using_modal_dialogs/) which is worth a look. This would be creating an in-page modal instead of the confirmation dialog, and perhaps it's easier to test. Whichever solution you go with, is good for now. 

(@KatiRG the above recipe mentions using `outlet somecontroller`. wondering if this would be the solution for rendering nested resources using their respective controllers, without a transition)

As for the feature request, yes, and if you can't label it I guess it's a repo owner thing, I'll add it myself.

**Comment by Qianfinland** [comment]
thanks for the link:) haha, another option to play with

**Comment by Qianfinland** [comment]
could you merge the code which fix the bug and test?I fix the bug already and want ro make a new feature request on the pop-up alert when users click "Remove support for this initiative".

File: tests/acceptance/making-a-suggestion-to-initiative-test.js [hunk 2]
```diff
@@ -31,3 +31,18 @@ test('Successfully', function(){
     });     
   });
 });
+
+// test on making suggestion without support this initiative by manully typing url to suggestion new page
```
Comment (oliverbarnes) [inline tests/acceptance/making-a-suggestion-to-initiative-test.js]: The scenario described in the comment here is what you want on the test description

File: tests/acceptance/making-a-suggestion-to-initiative-test.js [hunk 3]
```diff
@@ -31,3 +31,18 @@ test('Successfully', function(){
...
+// this shouldn't happen as only who support the initiative can make suggestion for it
+test('Manually typing url to make suggestion', function(){
+  visit('/initiatives/1/suggestions/new').then(function() {
+    expect( $("a:contains('Support this initiative')").length ).to.equal(1);
```
Comment (oliverbarnes) [inline tests/acceptance/making-a-suggestion-to-initiative-test.js]: if the 'Support this initiative' link is there, does it mean it's not supported yet? If so, the steps after this shouldn't work (a user shouldn't be able to submit a suggestion), right?

File: tests/acceptance/supporting-an-initiative-test.js [hunk 4]
```diff
@@ -19,10 +19,56 @@ test('Successfully', function(){
...
+  visit('/initiatives/1').then(function() {
+    click( $("a:contains('Support this initiative')") ).then(function() {
+      expect(currentURL()).to.equal('/initiatives/1');
+      expect(find('.support').text()).to.equal('Remove support for this initiative');
+      expect(find('.new_suggestion').text()).to.equal('Make a suggestion');
+      click( $("a:contains('Make a suggestion')") ).then(function() {
+        expect(currentURL()).to.equal('/initiatives/1/suggestions/new');
+        click( $("a:contains('Remove support for this initiative')") ).then(function() {
+          expect(currentURL()).to.equal('/initiatives/1');
+          expect(find('.support').text()).to.equal('Support this initiative');
```
Comment (oliverbarnes) [inline tests/acceptance/supporting-an-initiative-test.js]: this is a good scenario to test, but right now it's incomplete - it doesn't go as far as testing that a suggestion can't be made.

File: tests/acceptance/supporting-an-initiative-test.js [hunk 5]
```diff
@@ -19,10 +19,56 @@ test('Successfully', function(){
...
       });
     });
   });
 });
+
+
+test('Remove support after making suggestion', function(){
+  visit('/initiatives/1').then(function() {
+    click( $("a:contains('Support this initiative')") ).then(function() {
+      expect(currentURL()).to.equal('/initiatives/1');
```
Comment (oliverbarnes) [inline tests/acceptance/supporting-an-initiative-test.js]: no need to test the currentURL here, it's already in the main success scenario. In more specific scenarios like this one, you can focus on just on the different behavior for it.

File: tests/acceptance/supporting-an-initiative-test.js [hunk 6]
```diff
@@ -19,10 +19,56 @@ test('Successfully', function(){
...
+        expect(currentURL()).to.equal('/initiatives/1/suggestions/new');
+        expect($(':submit').attr('disabled')).to.equal('disabled');
+        fillIn('div.details textarea', "Make sure there's a doctor available for house calls");
+        click('form input[type=submit]').then(function() {
+          expect( $("a:contains('Remove support for this initiative')").length ).to.equal(1);
+          expect(currentURL()).to.equal('/initiatives/1');
+          expect(find('.suggestion').first().text()).to.equal("Make sure there's a doctor available for house calls");
+          click( $("a:contains('Remove support for this initiative')") ).then(function() {
+            expect(currentURL()).to.equal('/initiatives/1');
+            expect(find('.support').text()).to.equal('Support this initiative');
```
Comment (oliverbarnes) [inline tests/acceptance/supporting-an-initiative-test.js]: again, not going all the way and testing what should happen to the suggestion once the the support is removed.

**Comment by oliverbarnes** [comment]
Sorry, I don't see how this can be merged yet - how is the bug fixed and tested?

File: tests/acceptance/making-a-suggestion-to-initiative-test.js (hunk 3 above)
Comment (Qianfinland) [inline tests/acceptance/making-a-suggestion-to-initiative-test.js]: yes,normally they cannot make suggestion without supporting, should i modify code that they shouldn't make suggestion and let the test break?

File: tests/acceptance/supporting-an-initiative-test.js (hunk 4 above)
Comment (Qianfinland) [inline tests/acceptance/supporting-an-initiative-test.js]: what you mean "right is incomplete - it doesn't go as far as testing that a suggestion can't be made"

File: tests/acceptance/supporting-an-initiative-test.js [hunk 7]
```diff
@@ -19,10 +19,56 @@ test('Successfully', function(){
...
+      expect(find('.new_suggestion').text()).to.equal('Make a suggestion');
+      click( $("a:contains('Make a suggestion')") ).then(function() {
+        expect(currentURL()).to.equal('/initiatives/1/suggestions/new');
+        expect($(':submit').attr('disabled')).to.equal('disabled');
+        fillIn('div.details textarea', "Make sure there's a doctor available for house calls");
+        click('form input[type=submit]').then(function() {
+          expect( $("a:contains('Remove support for this initiative')").length ).to.equal(1);
+          expect(currentURL()).to.equal('/initiatives/1');
+          expect(find('.suggestion').first().text()).to.equal("Make sure there's a doctor available for house calls");
+          click( $("a:contains('Remove support for this initiative')") ).then(function() {
```
Comment (Qianfinland) [inline tests/acceptance/supporting-an-initiative-test.js]: so i should the delete whole code section on click( $("a:contains('Remove support for this initiative')") ).then(function() {}?

**Comment by Qianfinland** [comment]
On Mon, Sep 15, 2014 at 4:50 PM, Oliver Azevedo Barnes <
notifications@github.com> wrote:

> Sorry, I don't see how this can be merged yet - how is the bug fixed and
> tested?

the bug here is when you click "make a suggestion" then click "remove support.." the suggestion box and submit button is still there. I fix it by add this.transitionTo('initiative'); in the removeSupport function under routes/initiative.js. Add test code on during make suggestion and after finishing making suggestion if you click remove support.

File: tests/acceptance/making-a-suggestion-to-initiative-test.js (hunk 3 above)
Comment (oliverbarnes) [inline tests/acceptance/making-a-suggestion-to-initiative-test.js]: No, I mean the test should describe the expected behavior, which is to not be able to make a suggestion. As it is, you have one expectation for a a disabled submit button:

```
expect($(':submit').attr('disabled')).to.equal('disabled');
```

But then click on the submit button anyway:

```
click('form input[type=submit]').then(function() {
```

with an expectation for a suggestion to be created:

```
expect(find('.suggestion').first().text()).to.equal("Make sure there's a doctor available for house calls");
```

Which is the opposite of what it should do

File: tests/acceptance/supporting-an-initiative-test.js (hunk 4 above)
Comment (oliverbarnes) [inline tests/acceptance/supporting-an-initiative-test.js]: "right now it's incomplete" ;)

**Comment by oliverbarnes** [comment]
Ok, got it. So this PR should be downsized to just this fix and the test describing it clearly. Any expectations that are duplicate from the other scenarios should be removed. Any test that's incomplete (say, work in progress for the popup functionality), should be removed and added to the future PR dealing with the new functionality.

If this is confusing, maybe Miika can help you with it?

File: tests/acceptance/making-a-suggestion-to-initiative-test.js (hunk 3 above)
Comment (Qianfinland) [inline tests/acceptance/making-a-suggestion-to-initiative-test.js]: - ah, the code i write is to describe the unmormal thing happen here, i didn't write the expected way as i am not sure how users make suggestions via manually typing the url. What do you think how they make suggestion via the manual url?

-when users manually type the url, the suggestion box and submit button are there, and the submit button will  be disbled. When"support this initiative" is clicked, the submit button is abled?

-or when they maually typing the url, the suggestion box and submit button will not appear. when click the "support this initiative" they will show up?

-or anther way?

**Comment by Qianfinland** [comment]
yes, i think this PR is just for fix bug on when user click "Remove support
for this initaitive" and add test for removing support during making
suggestion and after finishing suggestion.

i will remove the test on making suggestion on manually typing url
i will also remove the pop-up alert code

**Comment by Qianfinland** [comment]
-i remove test code on making suggestion via manually typing url 
-i remove the pop-up alert code
-i remove duplicated code in the test , i keep less code after user click "Remove support for this initiative"
waiting for feedback and merge :) thanks

**Comment by oliverbarnes** [comment]
This is getting better. But what I'd like to see is a clear test that ensure the bug doesn't exist anymore. So if the bug is

> when you click "make a suggestion" then click "remove support.." the suggestion box and submit button is still there.

The test should go through these steps and the expect that the suggestion box and submit button _aren't_ there

**Comment by Qianfinland** [comment]
yes, the bug like that. Thanks for your feedback, before i focus on testing
when removing support user can support. Now i add test on suggestion box
and submit button are not there after removing support.

File: tests/acceptance/supporting-an-initiative-test.js [hunk 8]
```diff
@@ -19,10 +19,49 @@ test('Successfully', function(){
     click( $("a:contains('Support this initiative')") ).then(function() {
       expect(currentURL()).to.equal('/initiatives/1');
       expect(find('.support').text()).to.equal('Remove support for this initiative');
+      expect(find('.new_suggestion').text()).to.equal('Make a suggestion');
+      // test on removing support before making suggestion
```
Comment (oliverbarnes) [inline tests/acceptance/supporting-an-initiative-test.js]: This is the success scenario for supporting an initiative, so testing removal of support here is out of scope. Notice how the comment here is really describing another scenario?

I think this test suite can be renamed to a general 'Initiative support'

```
suite('Initiative Support', {
```

And then have just two scenarios, 'Adding support' and 'Removing support'. Ensuring that the suggestion text field is not present after removing support is enough

Comment (Qianfinland) [inline tests/acceptance/supporting-an-initiative-test.js]: for the Removing support, i don't need to divided removing support into
before/during/after making suggesting?

File: tests/acceptance/supporting-an-initiative-test.js [hunk 9]
```diff
@@ -14,15 +14,29 @@ suite('Supporting an initiative', {
...
+    });
+  });
+});
+
+
+test('Removing support ', function(){
+  visit('/initiatives/1').then(function() {
+    click( $("a:contains('Support this initiative')") ).then(function() {
+      click( $("a:contains('Make a suggestion')") ).then(function() {
+        expect(currentURL()).to.equal('/initiatives/1/suggestions/new');
```
Comment (oliverbarnes) [inline tests/acceptance/supporting-an-initiative-test.js]: again, this currentURL() expectation is already tested on the Making a Suggestion tests. This one and the next expectation for the disabled button are redundant, you can remove them and focus on just this scenario's specific expectations.

File: tests/acceptance/supporting-an-initiative-test.js [hunk 10]
```diff
@@ -14,15 +14,29 @@ suite('Supporting an initiative', {
...
+        expect($(':submit').attr('disabled')).to.equal('disabled');
+        click( $("a:contains('Remove support for this initiative')") ).then(function() {
+          expect($("textarea").length).to.equal(0);
```
Comment (oliverbarnes) [inline tests/acceptance/supporting-an-initiative-test.js:33]: also `expect(currentURL()).to.equal('/initiatives/1')` , and Make a suggestion should not be visible

File: app/routes/initiative.js [hunk 11]
```diff
@@ -6,11 +6,6 @@ var InitiativeRoute = Ember.Route.extend({
     return this.store.find('initiative', params.initiative_id);
   },
 
-  setupController: function(controller, model) {
```
Comment (Qianfinland) [inline app/routes/initiative.js:9]: this setupController is from master, we don't use it so i delete it

**Comment by Qianfinland** [comment]
modification is done, wating feedback

**Comment by oliverbarnes** [comment]
:+1: applied the patch locally and pushed to master

**Comment by oliverbarnes** [comment]
see [updated commit](https://github.com/oliverbarnes/participate-frontend/commit/6a490548e9cf21e3ca3dcbdf860bcd8689d644b2). I changed the commit message and changelog entry to be more descriptive

**Comment by KatiRG** [comment]
(@KatiRG the above recipe mentions using outlet somecontroller. wondering if this would be the solution for rendering nested resources using their respective controllers, without a transition)

it still seems that you would need to click something, if i understand the description correctly?

**Comment by oliverbarnes** [comment]
Looks like you can [use renderTemplate with specific outlets](http://emberjs.com/guides/routing/rendering-a-template/) for this, avoiding the click and transition

**Comment by oliverbarnes** [comment]
Also, interesting [blog post on ember outlets](https://blog.safaribooksonline.com/2013/09/10/ember-js-outlets/)

**Comment by oliverbarnes** [comment]
...the post is a good read to understand the generation of implicit routes and controllers, and how they wire the templates

**Pull request closed without merging**